pkg/identity/buildkite/principal.go

```diff
@@ -69,16 +111,23 @@ func (p jobPrincipal) Name(_ context.Context) string {
 }
 
 func (p jobPrincipal) Embed(_ context.Context, cert *x509.Certificate) error {
-	// Set SubjectAlternativeName to the pipeline URL on the certificate
-	parsed, err := url.Parse(p.url)
+	baseURL, err := url.Parse(p.url)
 	if err != nil {
 		return err
 	}
-	cert.URIs = []*url.URL{parsed}
+
+	pipelineUrl := baseURL.JoinPath(p.organizationSlug, p.pipelineSlug)
+	jobUrl := baseURL.JoinPath(p.organizationSlug, p.pipelineSlug, "builds", strconv.FormatInt(p.buildNumber, 10)+"#"+p.jobId)
+
+	// Set SubjectAlternativeName to the pipeline URL on the certificate
+	cert.URIs = []*url.URL{pipelineUrl}
 
 	// Embed additional information into custom extensions
 	cert.ExtraExtensions, err = certificate.Extensions{
-		Issuer: p.issuer,
+		Issuer:                 p.issuer,
+		RunInvocationURI:       jobUrl.String(),
+		RunnerEnvironment:      p.runnerEnvironment,
+		SourceRepositoryDigest: p.buildCommit,
```

Comment:
I would love to see support for a few additional extensions so we could get richer provenance info:
- Would it not be possible to also forward the repo url and ref from the originating source host to populate `SourceRepositoryURI` and `SourceRepositoryRef`? What about the repo and owner IDs? 
- Could you point the `BuildConfigURI` to [`pipeline.yml`](https://buildkite.com/docs/pipelines/defining-steps) file that configures the build? Is there something else that configures the build? Is the sha of this always the same as `p.buildCommit`?
- Is there not some concept of the event trigger that kicks of the job in buildkite that could be used to populate `BuildTrigger`?

Reply:
Buildkite has no access to the source repository.

Because of this, `SourceRepositoryURI` and `SourceRepositoryRef` are resolved within the runner environment, and are not always known on the Buildkite side which generates the OIDC tokens. We have a `repository` url which is provided as an input to the runner, but it is not always a fully-qualified URI. We have a build `branch` and `commit`, but these often start as values like `HEAD` and are resolved to a concrete `ref` by the runner which sends back the resolve commit sha.

Similarly for `BuildConfigURI`. The initial steps are defined as part of the pipeline on Buildkite. But those initial steps are usually just a single "upload" step. That step runs within the runner environment, to access the source repository. It conventionally looks at `buildkite.yml` or `.buildkite/pipeline.yml`, but can choose any file in the source repository. It may not even be a file, the runner may generate content, like with a script, which is piped into an upload command to use as build config. And the build is not limited to one upload - many files may be uploaded - and in fact the build config may be piped in and not sourced from an actual file in the repository.

`BuildTrigger` is probably something we could add. But I'd lean towards another PR for that.

Reply:
(I'd love to see the contents of this PR land as an increment, and consider additional extensions later.)